aea/cli/upgrade.py

```diff
@@ -103,23 +110,42 @@ def upgrade_project(ctx: Context) -> None:  # pylint: disable=unused-argument
     """Perform project upgrade."""
     click.echo("Starting project upgrade...")
 
+    interactive = ctx.config.get("interactive", True)
     old_component_ids = ctx.agent_config.package_dependencies
     item_remover = ItemRemoveHelper(ctx, ignore_non_vendor=True)
     agent_items = item_remover.get_agent_dependencies_with_reverse_dependencies()
     items_to_upgrade = set()
-    upgraders: List[ItemUpgrader] = []
+    upgraders: Dict[PackageId, ItemUpgrader] = {}
     shared_deps: Set[PackageId] = set()
     shared_deps_to_remove = set()
     items_to_upgrade_dependencies = set()
 
     # update aea_version in case current framework version is
     version = Version(aea.__version__)
     if not ctx.agent_config.aea_version_specifiers.contains(version):
+        old_aea_version_specifier = ctx.agent_config.aea_version
         new_aea_version_specifier = compute_specifier_from_version(version)
+        click.echo(
+            f"Upgrading version specifier from {old_aea_version_specifier} to {new_aea_version_specifier}."
+        )
         ctx.agent_config.aea_version = new_aea_version_specifier
         ctx.dump_agent_config()
 
+    eject_helper = InteractiveEjectHelper(ctx, agent_items, interactive=interactive)
+    eject_helper.get_latest_versions()
+    if len(eject_helper.item_to_new_version) == 0:
+        click.echo("Everything is already up to date!")
+        return
+    if not eject_helper.can_eject():
+        click.echo("Abort.")
+        return
+    eject_helper.eject()
+
     for package_id, deps in agent_items.items():
+        if package_id in eject_helper.to_eject:
+            continue
```

Comment:
can we isolate some code blocks here and give them proper names; for instance starting at 144, what is the code doing? would be much easier to read if this was better structured